client/src/components/categorical/value/index.js

```diff
@@ -171,16 +192,24 @@ class CategoryValue extends React.Component {
     If and only if true, update the component
     */
     const { props, state } = this;
-    const { metadataField, categoryIndex, categoricalSelection } = props;
-    const { categoricalSelection: newCategoricalSelection } = nextProps;
-
+    const {
+      metadataField,
+      categoryIndex,
+      categoricalSelection,
+      categorySummary,
+    } = props;
+    const {
+      categoryIndex: newCategoryIndex,
+      categoricalSelection: newCategoricalSelection,
+      categorySummary: newCategorySummary,
+    } = nextProps;
+
+    const label = categorySummary.categoryValues[categoryIndex];
+    const newLabel = newCategorySummary.categoryValues[newCategoryIndex];
+    const labelChanged = label !== newLabel;
     const valueSelectionChange =
-      categoricalSelection[metadataField].categoryValueSelected[
-        categoryIndex
-      ] !==
-      newCategoricalSelection[metadataField].categoryValueSelected[
-        categoryIndex
-      ];
+      (categoricalSelection[metadataField].get(label) ?? true) !==
```

Comment:
for my education, what edge case does `??` help take care of?

I'm imagining `categoricalSelection[metadataField]` could be either `undefined` or a value, so just wondering if `categoricalSelection[metadataField].get(label) !== newCategoricalSelection[metadataField].get(newLabel)` would do?

Reply:
Good catch @tihuan, I would agree. @bkmartinjr was there a corner case here?

Reply:
the corner case is when the reducer does not (yet) know about the selection state (because data is being loaded or otherwise transformed).  In this case, our selection state defaults to true, and if you change the code as proposed, it would eval incorrectly (it changes the default to false).

Reply:
Synced with Bruce via Slack!

Gonna try to summarize my understanding here:

Inside the rendering logic, we default the selection state of a label to `true`, if a user hasn't touched the label yet. AKA: `if (categoricalSelection[metadataField].get(label)) === undefined) return true`

So inside `shouldComponentUpdate()` here, `const valueSelectionChange` needs to check if the selection state of a label has changed from `true` to `false`, or `false` to `true`.

But since `categoricalSelection[metadataField].get(label))` could be `undefined`, we need to coalesce `undefined` to `true`, so when in async model, we don't trigger an unnecessary re-render when `categoricalSelection` state gets populated asynchronously:

E.g.,

Phase 1: `categoricalSelection` is NOT populated with persisted data yet:

```ts
shouldComponentUpdate() {
  const oldSelection = categoricalSelection[metadataField]?.get(label) // undefined
  const newSelection = newCategoricalSelection[metadataField]?.get(newLabel) // undefined

  const valueSelectionChange = oldSelection !== newSelection // false, since undefined === undefined
}
```

Phase 2: `categoricalSelection` is now asynchronously loaded with the persisted data:

```ts
shouldComponentUpdate() {
  const oldSelection = categoricalSelection[metadataField]?.get(label) // undefined
  // Turned out that `newLabel` is actually selected according to the persisted data, instead of unknown
  const newSelection = newCategoricalSelection[metadataField]?.get(newLabel) // true

  // This causes unnecessary re-render, because `undefined` should be coalesced to `true`
  const valueSelectionChange = oldSelection !== newSelection // true, since undefined !== true.
}
```

So I'm assuming when we have the async model, there's a case that `categoricalSelection` could be asynchronously populated with `true` as a default value for a category's label?

Because if a category in `categoricalSelection` is **always** populated with `new Map()`, then by default, `undefined` gets coalesced to `true`, so all labels are selected. And therefore a label's selection needs to transition from `undefined` to `false` in `categoricalSelection`, **before** it can transition to `true`?

For example,

Step 1: 
User sees all labels are selected
`categoricalSelection[aCategory].get(aLabel)`: `undefined`

Step 2:
User deselects label: `aLabel`
`categoricalSelection[aCategory].get(aLabel)`: `false`

Step 3:
User reselects label: `aLabel`
`categoricalSelection[aCategory].get(aLabel)`: `true`

And there doesn't exist a state transition of a label's selection from `undefined` to `true`, because `undefined` is `true` already, so the Phase 2 comparison scenario will never happen, and therefore:

```ts
const valueSelectionChange =
  categoricalSelection[metadataField].get(label) !== newCategoricalSelection[metadataField].get(newLabel)
```

would work?

I think I'm really close to understanding the reason behind using `?? true` here now lol 

Thanks for your patience, everyone 😂

Reply:
resolved offline.  fix pushed.